### PR TITLE
For transition profiler callbacks, always load the thread

### LINKED ISSUE #105070: Profiling problem on armv7l
Hello!

I've got a segmentation fault while profiling Tizen Xamarin-based Application with

a profiler based on SampleProfiler using ICorProfiler API:

https://github.com/dotnet/samples/tree/main/core/profiling

The segmentation fault always occurs here:

https://github.com/dotnet/runtime/blob/8602bb3b4003dbf5187385f58b29bd87e9468670/src/coreclr/vm/stubhelpers.cpp#L550

> Thread 1 "Alarm.dll" received signal SIGSEGV, Segmentation fault.
> 
> StubHelpers::ProfilerBeginTransitionCallback (pSecretParam=3040227328, pThread=0x0, unsafe_pThis=<optimized out>)
> 
>     at /usr/src/debug/coreclr-8.0.0-0.arm/src/coreclr/vm/stubhelpers.cpp:555
> 
> 555                GCX_PREEMP_THREAD_EXISTS(pThread);
> 
> (gdb) bt
> 
> `#0`  StubHelpers::ProfilerBeginTransitionCallback (pSecretParam=3040227328, pThread=0x0, unsafe_pThis=<optimized out>)
> 
>     at /usr/src/debug/coreclr-8.0.0-0.arm/src/coreclr/vm/stubhelpers.cpp:555
> 
> `#1`  0xb49836ec in ?? ()
> 
> Backtrace stopped: previous frame identical to this frame (corrupt stack?)
> 
> (gdb)
> 

This can be reproduced on armel (armv7l) machine on release/7.0 and release/8.0 branches, not release/6.0

So, I've found that PR #69761 eliminates the null-checking code for some reason:

https://github.com/dotnet/runtime/pull/69761/files#diff-c10abbd9bedf89c7a6ea5ace199b19ac17564e0e1e8583cd22b05546d18e0789L658-L659:~:text=%7D-,GCX_PREEMP_THREAD_EXISTS,-(pThread)%3B

Can you please describe why that change was made and why segfault can happen now?

cc @dotnet/samsung

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Also, add tests for profiler callbacks on a new thread to the runtime

Fixes #105070

### CODE REVIEW

**Review by davmason — APPROVED** [review]
Thanks for the fix, I don't know the details of the interop space enough to comment on the fix but the profiler tests look good to me

File: src/coreclr/vm/dllimport.cpp [hunk 1]
```diff
@@ -2272,15 +2272,7 @@ DWORD NDirectStubLinker::EmitProfilerBeginTransitionCallback(ILCodeStream* pcsEm
...
-    if (SF_IsForwardStub(dwStubFlags))
-    {
-        pcsEmit->EmitLDLOC(GetThreadLocalNum());
-    }
-    else
-    {
-        // we use a null pThread to indicate reverse interop
-        pcsEmit->EmitLoadNullPtr();
-    }
+    pcsEmit->EmitLDLOC(GetThreadLocalNum());
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:2275]: Will we always have a thread local? I like the explicit load of `null`, but I'm fine if we are guaranteed to always have a `null` initialized local. That means we also need to be clear that we aren't using `SkipLocalsInit` for the stub or its module.

Comment (jkoritzinsky) [reply src/coreclr/vm/dllimport.cpp:2275]: Yes, the thread local is created (and IL emitted to initialize it) on command when `GetThreadLocalNum()` is called.

**Review by mdh1418 — COMMENTED** [review]
I don't fully understand all the IL code emitting details, but from https://github.com/dotnet/runtime/pull/69761, the change to `EmitProfilerEndTransitionCallback` to have `_ASSERTE(SF_IsForwardStub(dwStubFlags));` + defaulting to `!fReverseInterop` code paths in stubhelpers.cpp seems to suggest that Profiler transitions shouldn't occur during reverse interops, but even if they did, behave the same as forward interops? Am I understanding that correctly?

Could we add asserts that pThread is not null in stubhelpers.cpp ProfilerBeginTransitionCallback/ProfilerEndTransitionCallback?

**Comment by jkoritzinsky** [comment]
> I don't fully understand all the IL code emitting details, but from #69761, the change to `EmitProfilerEndTransitionCallback` to have `_ASSERTE(SF_IsForwardStub(dwStubFlags));` + defaulting to `!fReverseInterop` code paths in stubhelpers.cpp seems to suggest that Profiler transitions shouldn't occur during reverse interops, but even if they did, behave the same as forward interops? Am I understanding that correctly?
> 
> Could we add asserts that pThread is not null in stubhelpers.cpp ProfilerBeginTransitionCallback/ProfilerEndTransitionCallback?

The profiler transitions for "reverse interop" are done as part of the JIT code emit now for these scenarios, so we don't handle them in the IL stub (which is what #69761 was fixing).

I can add the assert.

**Review by mdh1418 — APPROVED** [review]
Looks good to me!

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by gbalykov** [comment]
@jkoritzinsky thanks! Does this need to be backported to .net 7 and .net 8?

**Comment by jkoritzinsky** [comment]
I'll backport this to .NET 8. 7 is EOL so I can't backport there.

**Pull request merged**